src/lib/hooks/useBlobMemoryManagement.ts

```diff
@@ -0,0 +1,151 @@
+import { useEffect, useState, useRef } from 'react';
+import { useMatrixImage } from './useMatrixImage';
+import { queryClient } from '../web3/rainbowkit/provider';
+
+interface UseBlobUrlMemoryManagementOptions {
+  rootMargin?: string;
+  threshold?: number;
+  isThumbnail?: boolean;
+}
+
+/**
+ * A hook that tracks image visibility in the viewport
+ * - Monitors when images enter and exit the viewport
+ * - Works with existing blob URLs created elsewhere in the application
+ * - Handles Matrix URLs through the useMatrixImage hook
+ *
+ * @param url The image URL (can be a regular URL, data URL, blob URL, or Matrix URL)
+ * @param options Configuration options
+ * @returns The element ref to attach to your component and the URL to use
+ */
+export function useBlobUrlMemoryManagement<T extends HTMLElement = HTMLDivElement>(
+  url: string | undefined,
+  options: UseBlobUrlMemoryManagementOptions = {}
+) {
+  const { rootMargin = '100px', threshold = 0.1, isThumbnail = false } = options;
+
+  const [isInView, setIsInView] = useState(false);
+  const elementRef = useRef<T | null>(null);
+  const observerRef = useRef<IntersectionObserver | null>(null);
+  const timeoutRef = useRef<number | null>(null);
+  const [isUrlRevoked, setIsUrlRevoked] = useState(false);
+  const originalUrlRef = useRef<string | undefined>(url);
+
+  useEffect(() => {
+    originalUrlRef.current = url;
+  }, [url]);
+
+  // Use the Matrix image hook for Matrix URLs
+  const { data: effectiveUrl, isPending: isMatrixLoading } = useMatrixImage(url, {
```

Comment:
isPending is recommended over isLoading for most cases
https://github.com/TanStack/query/discussions/6297#discussioncomment-7467010